src/compiler/emitter.ts

```diff
@@ -144,81 +147,193 @@ module ts {
+                for (let node = container; isNodeDescendentOf(node, container); node = node.nextContainer) {
+                    if (node.locals && hasProperty(node.locals, name)) {
+                        // We conservatively include alias symbols to cover cases where they're emitted as locals
+                        if (node.locals[name].flags & (SymbolFlags.Value | SymbolFlags.ExportValue | SymbolFlags.Alias)) {
+                            return false;
+                        }
+                    }
                 }
-                else {
-                    currentScopeNames = undefined;
+                return true;
+            }
+
+            function generateNameForNode(node: Node): string {
+                switch (node.kind) {
+                    case SyntaxKind.FunctionDeclaration:
+                    case SyntaxKind.ClassDeclaration:
+                        return generateNameForFunctionOrClassDeclaration(<Declaration>node);
+                    case SyntaxKind.ModuleDeclaration:
+                        return generateNameForModuleOrEnum(<ModuleDeclaration>node);
+                    case SyntaxKind.EnumDeclaration:
+                        return generateNameForModuleOrEnum(<EnumDeclaration>node);
+                    case SyntaxKind.ImportDeclaration:
+                        return generateNameForImportDeclaration(<ImportDeclaration>node);
+                    case SyntaxKind.ExportDeclaration:
+                        return generateNameForExportDeclaration(<ExportDeclaration>node);
+                    case SyntaxKind.ExportAssignment:
+                        return generateNameForExportAssignment(<ExportAssignment>node);
+                    default: 
+                        Debug.fail("Unexpected node kind:" + node.kind);
+                        return undefined;
                 }
             }
 
-            function generateUniqueNameForLocation(location: Node, baseName: string): string {
-                let name: string
-                // first try to check if base name can be used as is
-                if (!isExistingName(location, baseName)) {
-                    name = baseName;
+            function isUniqueName(name: string) {
+                return !resolver.hasGlobalName(name) &&
+                    !hasProperty(currentSourceFile.identifiers, name) &&
+                    (!generatedNames || !hasProperty(generatedNames, name));
+            }
+
+            function isUniqueTempVariableName(location: Node, name: string): boolean {
+                if (!resolver.isUnknownIdentifier(location, name)) {
+                    return false;
                 }
-                else {
-                    name = generateUniqueName(baseName, n => isExistingName(location, n));
+
+                return !currentScopeTemp.names || !hasProperty(currentScopeTemp.names, name);
+            }
+
+            function makeUniqueName(baseName: string): string {

```

Comment:
Minor nit.  But once we try _name, then from that point on, i wouldn't prefix with _.   So it would be  'name', then '_name', then 'name_1', 'name_2', etc.  Adding a prefix+suffix just seems unnecessary.